src/core/qgsvectorlayer.cpp

```diff
@@ -2387,6 +2387,19 @@ long QgsVectorLayer::featureCount() const
          ( mEditBuffer ? mEditBuffer->mAddedFeatures.size() - mEditBuffer->mDeletedFeatureIds.size() : 0 );
 }
 
+long QgsVectorLayer::featureCount( QgsFeatureRequest request )
```

Comment:
In general if one is only interested in getting the feature count for a filtered set there is no need for an iterator which actually starts an operation and uses resources on the server and in QGIS. I assume that very often features will still be fetched though.